File: linkedList/deepCopy.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        #we're going to use a hash map! 
        #go through the list once and create the map
        #go through the map and fill in the map Nodes' next and random pointers
        hashMap = {None: None}
        cur = head
        while cur:
            hashMap[cur] = Node(cur.val)
            cur = cur.next
        
        cur = head
        while cur:
            hashMap[cur].next = hashMap[cur.next]
            hashMap[cur].random = hashMap[cur.random]
            cur = cur.next
        
        return hashMap[head]

File: linkedList/deepCopy.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        /*
        -USE A HASH MAP! (or a vector of pairs in this case?)
        - iterate thru LL once to create new nodes,
        - iterate again to set new nodes' pointers
        */
        unordered_map<Node*, Node*> hashMap;
        Node* tmp = head;
        while(tmp){
            Node* cpy = new Node(tmp -> val);
            hashMap[tmp] = cpy;
            tmp = tmp -> next;
        }
        tmp = head;
        while(tmp){
            hashMap[tmp] -> next = hashMap[tmp -> next];
            hashMap[tmp] -> random = hashMap[tmp -> random];
            tmp = tmp -> next;
        }
        return hashMap[head];
    }
};
